### PR TITLE
Dikenmaharjan10/saf 532 server route to retrieve manifest of what to encrypt for who

### PR DESCRIPTION
**Pull request by DikenMaharjan**

### CODE REVIEW

File: src/main/kotlin/com/safehill/SafehillClient.kt [hunk 1]
```diff
@@ -25,11 +27,26 @@ class SafehillClient private constructor(
     val webSocketApi: WebSocketApi,
     val currentUser: LocalUser
 ) {
+    val encryptionDetailsController by lazy {
```
Comment (gennaro-safehill) [inline src/main/kotlin/com/safehill/SafehillClient.kt:30]: What happens when the authentication state changes? Are these classes responsive to that event?

File: src/main/kotlin/com/safehill/kclient/controllers/ConversationThreadController.kt [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+                anchorId = threadId,
+                interactionAnchor = InteractionAnchor.THREAD
+            )
+            if (symmetricKey == null) {
+                throw UserInteractionController.InteractionErrors.MissingE2EEDetails(
+                    anchorId = threadId,
+                    anchor = InteractionAnchor.THREAD
+                )
+            } else {
+                serverProxy.provideEncryptionDetails(
```
Comment (gennaro-safehill) [inline src/main/kotlin/com/safehill/kclient/controllers/ConversationThreadController.kt]: Same comment as in the other PR. The verb "provide" has a specific meaning and we shouldn't use it in this context. Can you please align the SafehillAPI method names between Safehill-Client (Swift) and Safehill-kClient (Kotlin)?

File: src/main/kotlin/com/safehill/kclient/controllers/UserInteractionController.kt [hunk 3]
```diff
@@ -100,8 +100,9 @@ class UserInteractionController internal constructor(
         return if (existingThread != null) {
```
Comment (gennaro-safehill) [inline src/main/kotlin/com/safehill/kclient/controllers/UserInteractionController.kt:102]: Are we checking existingThread existence on remote?

File: src/main/kotlin/com/safehill/kclient/models/dtos/websockets/UserConversionManifestDTO.kt [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+@Serializable
+data class UserConversionManifestDTO(
+    val assetIdsByGroupId: AssetIdsByGroupId,
+    val newUser: RemoteUser,
+    val threadIds: List<String>
+) : WebSocketMessage
+
+
+//todo figure out the type of asset id by group id when dealing with asset and groups.
```
Comment (gennaro-safehill) [inline src/main/kotlin/com/safehill/kclient/models/dtos/websockets/UserConversionManifestDTO.kt:14]: By type you mean whether or not is a photo message?

Comment (gennaro-safehill) [reply src/main/kotlin/com/safehill/kclient/models/dtos/websockets/UserConversionManifestDTO.kt:14]: The asset descriptor would tell you that. 
You should get latest asset descriptors from remote server for the assets you have to send encryption details

File: src/main/kotlin/com/safehill/SafehillClient.kt (hunk 1 above)
Comment (DikenMaharjan) [reply src/main/kotlin/com/safehill/SafehillClient.kt:30]: When the authenticationState changes, the values in the currentUser will be mutated with the new values. So, yeah.. it is responsive to that event.
But tbh, I don't like the idea that the values in the LocalUser are mutable.

File: src/main/kotlin/com/safehill/kclient/controllers/ConversationThreadController.kt (hunk 2 above)
Comment (DikenMaharjan) [reply src/main/kotlin/com/safehill/kclient/controllers/ConversationThreadController.kt]: can you help me point to the file in Safehill-Client (Swift) that contains the functions?

File: src/main/kotlin/com/safehill/kclient/controllers/UserInteractionController.kt (hunk 3 above)
Comment (DikenMaharjan) [reply src/main/kotlin/com/safehill/kclient/controllers/UserInteractionController.kt:102]: yeah.. the serverProxy first checks the local server and then the remote server.

File: src/main/kotlin/com/safehill/kclient/models/dtos/websockets/UserConversionManifestDTO.kt (hunk 4 above)
Comment (DikenMaharjan) [reply src/main/kotlin/com/safehill/kclient/models/dtos/websockets/UserConversionManifestDTO.kt:14]: By type I was referring to the format of the json for the key "assetIdsByGroupId"... I didn't trigger the request to access to a group of a thread so I had no idea how the json would be formatted for these cases. I was planning to fix the type (model class corresponding to the json format) when I will be dealing with the group access or asset access.

File: src/main/kotlin/com/safehill/kclient/controllers/UserInteractionController.kt (hunk 3 above)
Comment (gennaro-safehill) [reply src/main/kotlin/com/safehill/kclient/controllers/UserInteractionController.kt:102]: Should it be the other way around? Check the remote, and fall back to local if there is no connection?
I believe that's what's happening in Safehill-Client

File: src/main/kotlin/com/safehill/SafehillClient.kt (hunk 1 above)
Comment (gennaro-safehill) [reply src/main/kotlin/com/safehill/SafehillClient.kt:30]: In Swift they are immutable but the reference to the user object in AppUserController changes

File: src/main/kotlin/com/safehill/kclient/controllers/UserInteractionController.kt (hunk 3 above)
Comment (DikenMaharjan) [reply src/main/kotlin/com/safehill/kclient/controllers/UserInteractionController.kt:102]: Yeah, I agree... Previously the thread's members could not be changed but now we should definitely check the remote server first. I had plans to talk about this in the meeting but forgot it.. I am glad that you caught it up.

**Comment by gennaro-safehill** [comment]
Can this be merged?

**Comment by DikenMaharjan** [comment]
> Can this be merged?
Yeah.. The refactoring is complete.

**Pull request merged**